Sources/Basics/VirtualFileSystem.swift

```diff
@@ -0,0 +1,225 @@
+        case .root(_): return "/"
+        }
+    }
+
+    var fileAttributeType: FileAttributeType {
+        switch self {
+        case .directory(_, _, _): return .typeDirectory
+        case .file(_, _, let isSymlink): return isSymlink ? .typeSymbolicLink : .typeRegular
+        case .root(_): return .typeDirectory
+        }
+    }
+
+    var isDirectory: Bool {
+        switch self {
+        case .directory(_, _, _): return true
+        case .file(_, _, _): return false
+        case .root(_): return true
+        }
+    }
+
+    var isFile: Bool {
+        switch self {
+        case .directory(_, _, _): return false
+        case .file(_, _, _): return true
+        case .root(_): return false
+        }
+    }
+
+    var isRoot: Bool {
+        switch self {
+        case .directory(_, _, _): return false
+        case .file(_, _, _): return false
+        case .root(_): return true
+        }
+    }
+
+    var isSymlink: Bool {
+        switch self {
+        case .directory(_, let isSymlink, _): return isSymlink
+        case .file(_, _, let isSymlink): return isSymlink
+        case .root(_): return false
+        }
+    }
+}
+
+private enum Errors: Swift.Error, LocalizedError {
+    case noSuchFileOrDirectory(path: AbsolutePath)
+    case notAFile(path: AbsolutePath)
+    case readOnlyFileSystem
+    case unhandledDirectoryNode(path: AbsolutePath)
+
+    public var errorDescription: String? {
+        switch self {
+        case .noSuchFileOrDirectory(let path): return "no such file or directory: \(path.pathString)"
+        case .notAFile(let path): return "not a file: \(path.pathString)"
+        case .readOnlyFileSystem: return "read-only filesystem"
+        case .unhandledDirectoryNode(let path): return "unhandled directory node: \(path.pathString)"
+        }
+    }
+}
+
+private extension FileSystem {
+    func getDirectoryNodes(_ path: AbsolutePath) throws -> [DirectoryNode] {
+        return try getDirectoryContents(path).compactMap {
+            let current = path.appending(component: $0)
+            let isSymlink = isSymlink(current)
+
+            if isFile(current) {
+                return .file(name: $0, isExecutable: isExecutableFile(current), isSymlink: isSymlink)
+            } else if isDirectory(current) {
+                if $0.hasPrefix(".") { return nil } // we ignore hidden files
+                return .directory(name: $0, isSymlink: isSymlink, children: try getDirectoryNodes(current))
+            } else {
+                throw Errors.unhandledDirectoryNode(path: current)
+            }
+        }
+    }
+}
+
+/// A JSON-backed, read-only virtual file system.
+public class VirtualFileSystem: FileSystem {
+    private let root: DirectoryNode
+    private let readFileContents: ((AbsolutePath) throws -> ByteString)
+
+    public init(path: AbsolutePath, fs: FileSystem, readFileContents: @escaping ((AbsolutePath) throws -> ByteString) = { _ in "" }) throws {
```

Comment:
readFileContents -> fileContentProvider

Reply:
also maybe instead embed the content in the JSON instead? probably base64 encoded etc

Reply:
Yah, that seems like a better idea. I changed it so that you pass in an array of paths for which to include contents as part of the JSON.